### PR TITLE
feature/#1568 update boolean apis to using reason collection

### PR DESCRIPTION
**Pull request by toan-quach**
#1568

### CODE REVIEW

File: taipy/core/_manager/_manager.py [hunk 1]
```diff
@@ -125,11 +126,16 @@ def _get(cls, entity: Union[str, EntityType], default=None) -> EntityType:
             return default
 
     @classmethod
-    def _exists(cls, entity_id: str) -> bool:
+    def _exists(cls, entity_id: str) -> ReasonCollection:
         """
         Returns True if the entity id exists.
```
Comment (trgiangdo) [inline taipy/core/_manager/_manager.py]: Do we also need to update the docstring for the return type?

The `is_ready_to_run()` method did get an update, but not the others

File: taipy/core/reason/reason.py [hunk 2]
```diff
@@ -155,3 +155,76 @@ class InvalidUploadFile(Reason, _DataNodeReasonMixin):
...
+class SubmissionStatusIsUndefined(Reason, _DataNodeReasonMixin):
+    """
+    The submission status is undefined.
+
+    Attributes:
+        submission_id (str): The submission identifier.
+    """
+
+    def __init__(self, submission_id: str):
+        Reason.__init__(self, f"The submission {submission_id} status is undefined.")
```
Comment (trgiangdo) [inline taipy/core/reason/reason.py]:
```suggestion
        Reason.__init__(self, f"The status of submission {submission_id} is undefined.")
```

File: taipy/core/_manager/_manager.py (hunk 1 above)
Comment (toan-quach) [reply taipy/core/_manager/_manager.py]: yep I believe we should, let me add that in

**Review by trgiangdo — APPROVED** [review]
LGTM! Thank you

File: taipy/core/reason/reason.py [hunk 3]
```diff
@@ -155,3 +155,76 @@ class InvalidUploadFile(Reason, _DataNodeReasonMixin):
     def __init__(self, file_name: str, datanode_id: str):
         Reason.__init__(self, f'The uploaded file {file_name} has invalid data for data node "{datanode_id}"')
         _DataNodeReasonMixin.__init__(self, datanode_id)
+
+
+class EntityDoesNotExist(Reason, _DataNodeReasonMixin):
```
Comment (jrobinAV) [inline taipy/core/reason/reason.py:160]: This Reason is not related to data nodes. It should not inherited from `_DataNodeReasonMixin`.

File: taipy/core/reason/reason.py [hunk 4]
```diff
@@ -155,3 +155,76 @@ class InvalidUploadFile(Reason, _DataNodeReasonMixin):
...
+
+    Attributes:
+        entity_id (str): The entity identifier.
+    """
+
+    def __init__(self, entity_id: str):
+        Reason.__init__(self, f"Entity {entity_id} does not exist in the repository.")
+
+
+class JobIsNotFinished(Reason, _DataNodeReasonMixin):
```
Comment (jrobinAV) [inline taipy/core/reason/reason.py:172]: This Reason is not related to data nodes. It should not inherited from `_DataNodeReasonMixin`.

File: taipy/core/reason/reason.py [hunk 5]
```diff
@@ -155,3 +155,76 @@ class InvalidUploadFile(Reason, _DataNodeReasonMixin):
...
+
+    Attributes:
+        job_id (str): The job identifier.
+    """
+
+    def __init__(self, job_id: str):
+        Reason.__init__(self, f"The job {job_id} is not finished yet.")
+
+
+class ScenarioIsThePrimaryScenario(Reason, _DataNodeReasonMixin):
```
Comment (jrobinAV) [inline taipy/core/reason/reason.py:184]: This Reason is not related to data nodes. It should not inherited from `_DataNodeReasonMixin`.

File: taipy/core/reason/reason.py [hunk 6]
```diff
@@ -155,3 +155,76 @@ class InvalidUploadFile(Reason, _DataNodeReasonMixin):
...
+    Attributes:
+        scenario_id (str): The scenario identifier.
+        cycle_id (str): The cycle identifier.
+    """
+
+    def __init__(self, scenario_id: str, cycle: str):
+        Reason.__init__(self, f"The scenario {scenario_id} is the primary scenario of cycle {cycle}.")
+
+
+class ScenarioDoesNotBelongToACycle(Reason, _DataNodeReasonMixin):
```
Comment (jrobinAV) [inline taipy/core/reason/reason.py:197]: This Reason is not related to data nodes. It should not inherited from `_DataNodeReasonMixin`.

File: taipy/core/reason/reason.py [hunk 7]
```diff
@@ -155,3 +155,76 @@ class InvalidUploadFile(Reason, _DataNodeReasonMixin):
...
+
+    Attributes:
+        scenario_id (str): The scenario identifier.
+    """
+
+    def __init__(self, scenario_id: str):
+        Reason.__init__(self, f"The scenario {scenario_id} does not belong to any cycle.")
+
+
+class SubmissionIsNotFinished(Reason, _DataNodeReasonMixin):
```
Comment (jrobinAV) [inline taipy/core/reason/reason.py:209]: This Reason is not related to data nodes. It should not inherited from `_DataNodeReasonMixin`.

File: taipy/core/reason/reason.py [hunk 8]
```diff
@@ -155,3 +155,76 @@ class InvalidUploadFile(Reason, _DataNodeReasonMixin):
...
+
+    Attributes:
+        submission_id (str): The submission identifier.
+    """
+
+    def __init__(self, submission_id: str):
+        Reason.__init__(self, f"The submission {submission_id} is not finished yet.")
+
+
+class SubmissionStatusIsUndefined(Reason, _DataNodeReasonMixin):
```
Comment (jrobinAV) [inline taipy/core/reason/reason.py:221]: This Reason is not related to data nodes. It should not inherited from `_DataNodeReasonMixin`.

**Comment by jrobinAV** [comment]
@toan-quach I believe we should remove the unnecessary inheritance. What do you think?

**Comment by toan-quach** [comment]
@jrobinAV you're right!!!! I missed that totally! Thanks for pointing it out!
